File: shared_space.py
"""Persistent per-user WinBridge Windows environment and Proton selection."""
from contextlib import contextmanager
import fcntl
import json
import os
from pathlib import Path
import tempfile


def data_dir():
    return Path(os.environ.get('XDG_DATA_HOME', Path.home() / '.local/share')) / 'winbridge'


def settings_path():
    return Path(os.environ.get('XDG_CONFIG_HOME', Path.home() / '.config')) / 'winbridge/settings.json'


def read_settings():
    try:
        settings = json.loads(settings_path().read_text())
    except FileNotFoundError:
        legacy = settings_path().parent.parent / 'protonrun/settings.json'
        try:
            settings = json.loads(legacy.read_text())
        except FileNotFoundError:
            return {}
    if not isinstance(settings, dict) or any(not isinstance(v, str) for v in settings.values()):
        raise ValueError('Ugyldige WinBridge-innstillinger.')
    return settings


def save_settings(settings):
    path = settings_path()
    path.parent.mkdir(parents=True, exist_ok=True)
    with tempfile.NamedTemporaryFile(mode='w', dir=path.parent, delete=False) as out:
        json.dump(settings, out, indent=2)
        temporary = Path(out.name)
    temporary.replace(path)


@contextmanager
def settings_lock():
    path = settings_path().with_suffix('.lock')
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open('a') as lock:
        fcntl.flock(lock, fcntl.LOCK_EX)
        yield


def shared_prefix(settings=None):
    settings = read_settings() if settings is None else settings
    if settings.get('prefix'):
        return Path(settings['prefix']).expanduser().resolve()
    # Adopt the sole legacy environment in place, preserving installed shortcuts and paths.
    old_data = data_dir().parent / 'protonrun'
    if (old_data / 'shared/pfx').is_dir():
        return (old_data / 'shared').resolve()
    old_prefixes = sorted(p.parent for p in (old_data / 'prefixes').glob('*/pfx') if p.is_dir())
    if len(old_prefixes) == 1:
        return old_prefixes[0].resolve()
    legacy = sorted(p.parent for p in (data_dir() / 'prefixes').glob('*/pfx') if p.is_dir())
    return legacy[0].resolve() if len(legacy) == 1 else (data_dir() / 'shared').resolve()


def select_proton(versions, chooser, validate, change=False):
    with settings_lock():
        settings = read_settings()
        saved = Path(settings['proton']) if settings.get('proton') else None
        if saved and (saved / 'proton').is_file() and not change:
            return saved
        if not versions:
            raise RuntimeError('Fant ingen Proton-versjoner. Installer Proton fra Verktøy i Steam.')
        selected = chooser(versions)
        if selected is None:
            return None
        validate(selected)
        settings.update(prefix=str(shared_prefix(settings)), proton=str(selected))
        save_settings(settings)
        return selected

File: src/shared_space.cpp
/*
    WinBridge v1.0
    Copyright (c) 2026 A. Blohmè <alexander.blohme@gmail.com>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://gnu.org>.
*/

#include "shared_space.h"
#include "launcher.h"

#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QJsonDocument>
#include <QJsonObject>
#include <QTemporaryFile>
#include <stdexcept>
#include <fcntl.h>
#include <sys/file.h>
#include <unistd.h>

namespace WinBridge {

QString dataDir() {
    QString base = qEnvironmentVariable("XDG_DATA_HOME");
    if (base.isEmpty()) {
        base = QDir::homePath() + "/.local/share";
    }
    return QDir::cleanPath(base + "/winbridge");
}

QString defaultPrefix() {
    return QDir::cleanPath(dataDir() + "/shared");
}

QString settingsPath() {
    QString conf = qEnvironmentVariable("XDG_CONFIG_HOME");
    if (conf.isEmpty()) {
        conf = QDir::homePath() + "/.config";
    }
    return QDir::cleanPath(conf + "/winbridge/settings.json");
}

QJsonObject readSettings() {
    QString path = settingsPath();
    QFile file(path);
    if (!file.exists()) {
        QDir confDir = QFileInfo(path).dir();
        confDir.cdUp();
        QString legacyPath = QDir::cleanPath(confDir.absolutePath() + "/protonrun/settings.json");
        file.setFileName(legacyPath);
        if (!file.exists()) {
            return QJsonObject();
        }
    }

    if (!file.open(QIODevice::ReadOnly)) {
        return QJsonObject();
    }

    QJsonParseError err;
    QJsonDocument doc = QJsonDocument::fromJson(file.readAll(), &err);
    if (err.error != QJsonParseError::NoError || !doc.isObject()) {
        throw std::runtime_error("Ugyldige WinBridge-innstillinger.");
    }

    QJsonObject obj = doc.object();
    for (auto it = obj.begin(); it != obj.end(); ++it) {
        if (!it.value().isString()) {
            throw std::runtime_error("Ugyldige WinBridge-innstillinger.");
        }
    }
    return obj;
}

bool saveSettings(const QJsonObject &settings) {
    QString path = settingsPath();
    QDir dir = QFileInfo(path).dir();
    if (!dir.exists()) {
        dir.mkpath(".");
    }

    QTemporaryFile temp(path + ".tmp.XXXXXX");
    if (!temp.open()) {
        return false;
    }

    QJsonDocument doc(settings);
    temp.write(doc.toJson(QJsonDocument::Indented));
    temp.flush();
    QString tempPath = temp.fileName();
    temp.close();

    QFile::remove(path);
    return QFile::rename(tempPath, path);
}

SettingsLock::SettingsLock() {
    QString path = settingsPath() + ".lock";
    QDir dir = QFileInfo(path).dir();
    if (!dir.exists()) {
        dir.mkpath(".");
    }
    fd = ::open(path.toUtf8().constData(), O_CREAT | O_RDWR, 0666);
    if (fd >= 0) {
        ::flock(fd, LOCK_EX);
    }
}

SettingsLock::~SettingsLock() {
    if (fd >= 0) {
        ::flock(fd, LOCK_UN);
        ::close(fd);
    }
}

QString sharedPrefix(const QJsonObject &passedSettings) {
    QJsonObject settings = passedSettings;
    if (settings.isEmpty()) {
        try {
            settings = readSettings();
        } catch (...) {
            settings = QJsonObject();
        }
    }

    if (settings.contains("prefix") && !settings["prefix"].toString().trimmed().isEmpty()) {
        QString p = settings["prefix"].toString().trimmed();
        if (p.startsWith("~/")) {
            p = QDir::homePath() + p.mid(1);
        }
        QFileInfo fi(p);
        return fi.canonicalFilePath().isEmpty() ? QDir::cleanPath(fi.absoluteFilePath()) : fi.canonicalFilePath();
    }

    QString oldData = QDir::cleanPath(dataDir() + "/../protonrun");
    if (QDir(oldData + "/shared/pfx").exists()) {
        QFileInfo fi(oldData + "/shared");
        return fi.canonicalFilePath().isEmpty() ? QDir::cleanPath(fi.absoluteFilePath()) : fi.canonicalFilePath();
    }

    QDir oldPrefixesDir(oldData + "/prefixes");
    if (oldPrefixesDir.exists()) {
        QStringList entries = oldPrefixesDir.entryList(QDir::Dirs | QDir::NoDotAndDotDot, QDir::Name);
        QStringList legacy;
        for (const QString &e : entries) {
            if (QDir(oldPrefixesDir.filePath(e + "/pfx")).exists()) {
                QFileInfo fi(oldPrefixesDir.filePath(e));
                legacy << (fi.canonicalFilePath().isEmpty() ? QDir::cleanPath(fi.absoluteFilePath()) : fi.canonicalFilePath());
            }
        }
        if (legacy.size() == 1) {
            return legacy[0];
        }
    }

    QDir curPrefixesDir(dataDir() + "/prefixes");
    if (curPrefixesDir.exists()) {
        QStringList entries = curPrefixesDir.entryList(QDir::Dirs | QDir::NoDotAndDotDot, QDir::Name);
        QStringList curLegacy;
        for (const QString &e : entries) {
            if (QDir(curPrefixesDir.filePath(e + "/pfx")).exists()) {
                QFileInfo fi(curPrefixesDir.filePath(e));
                curLegacy << (fi.canonicalFilePath().isEmpty() ? QDir::cleanPath(fi.absoluteFilePath()) : fi.canonicalFilePath());
            }
        }
        if (curLegacy.size() == 1) {
            return curLegacy[0];
        }
    }

    return defaultPrefix();
}

QString selectProton(
    const QStringList &versions,
    std::function<QString(const QStringList &)> chooser,
    std::function<void(const QString &)> validate,
    bool change
) {
    SettingsLock lock;
    QJsonObject settings;
    try {
        settings = readSettings();
    } catch (...) {
        settings = QJsonObject();
    }

    QString saved = settings.value("proton").toString();
    if (!saved.isEmpty() && isProtonAvailable(saved) && !change) {
        return saved;
    }

    if (versions.isEmpty()) {
        throw std::runtime_error("Fant ingen Proton-versjoner. Last ned Proton i innstillingene til WinBridge Manager.");
    }

    QString selected = chooser(versions);
    if (selected.isEmpty()) {
        return QString();
    }

    validate(selected);
    settings["prefix"] = sharedPrefix(settings);
    settings["proton"] = selected;
    saveSettings(settings);
    return selected;
}

} // namespace WinBridge
